R/utility.R

```diff
@@ -505,30 +473,73 @@ waitForTasksToComplete <- function(jobId, timeout) {
       currentTasks <-
         rAzureBatch::listTask(jobId, skipToken = URLdecode(skipTokenValue))
 
-      taskStates <-
-        lapply(currentTasks$value, function(x)
-          x$state != "completed")
+      totalTasks <- totalTasks + length(currentTasks$value)
+    }
 
-      for (i in 1:length(taskStates)) {
-        if (taskStates[[i]] == FALSE) {
-          count <- count + 1
+    pb <- txtProgressBar(min = 0, max = totalTasks, style = 3)
+    timeToTimeout <- Sys.time() + timeout
+
+    repeat {
+      taskCounts <- rAzureBatch::getJobTaskCounts(jobId)
+
+      setTxtProgressBar(pb, taskCounts$completed)
+
+
+      if (taskCounts$failed > 0 &&
+          errorHandling == "stop" &&
+          taskCounts$validationStatus == "Validated") {
+        cat("\n")
+
+        filter <- "state eq 'completed' and executionInfo/exitCode ne 0"
+        select <- "id, executionInfo"
+
+        failedTasks <-
+          rAzureBatch::listTask(jobId, filter = filter, select = select)
+
+        tasksFailureWarningLabel <-
+          sprintf("There are currently %i tasks failed while running the job:\n", taskCounts$failed)
+
+        for (i in 1:length(failedTasks$value)) {
+          if (failedTasks$value[[i]]$executionInfo$result == "Failure") {
+            tasksFailureWarningLabel <-
+              paste0(tasksFailureWarningLabel,
+                     sprintf("%s\n", failedTasks$value[[i]]$id))
+          }
         }
+
+        warning(sprintf(tasksFailureWarningLabel,
+                        taskCounts$failed))
+
+        response <- rAzureBatch::terminateJob(jobId)
+        httr::stop_for_status(response)
+
+        stop(sprintf(
+          paste("Errors have occurred while running the job '%s'.",
+                "Error handling is set to 'stop' and has proceeded to terminate the job.",
+                "The user will have to handle deleting the job.",
+                "If this is not the correct behavior, change the errorHandling property",
+                "in the foreach object. Use the 'getJobFile' function to obtain the logs.",
+                "For more information about getting job logs, follow this link:",
+                paste0("https://github.com/Azure/doAzureParallel/blob/master/docs/",
+                       "40-troubleshooting.md#viewing-files-directly-from-compute-node")),
+          jobId
+        ))
       }
-    }
 
-    setTxtProgressBar(pb, count)
+      if (Sys.time() > timeToTimeout) {
+        stop("A timeout has occurred when waiting for tasks to complete. ")
```

Comment:
Be more verbose around what the timeout was, and how a user can change it.

Reply:
The job will continue, the user will need to track it manually